### ISSUE TITLE
Consider to not use Extism

### ISSUE THREAD

**Issue by Snowiiii** [Snowiiii/Pumpkin#110]
The Plan was to use [Extism](https://crates.io/crates/extism) for Plugins, It has really promising features and enables us to "theoretically" support multiple programming languages for Plugins. Everything runs in WebAssembly which is a runtimes, This means it also has its own Overhead. But that is my smallest concern.

### Real Pros
- **Cross Platform**
- **Multi Language**

### The Problems:
- **Compile time/Indexing time**: Extism is really Heavy, mostly because of Wasm and other WebAssembly libraries. I mean such a runtime requires a good amount of code. Extism doubles our overall dependency count and makes our `Cargo.lock` so much bigger. You will begin to notice the real problem when compiling or if your code editor performs dependency indexing. I noticed it on my PC it was not good but not that bad, But when i working on my Laptop it is really not funny to work with.
- **Limited Language support**: So i said "theoretically" at the beginning when i told that it is possible to write Plugins in different Programming languages. That's because we still would have to write, maintain and publish API's for Pumpkin in every new Language (Which is very hard).
- **Performance**: Also like i already mentioned, WebAssembly is a runtime. A Runtime will always have more overhead than great optimized binary code, Im not sure if Plugins are really Performance critical, But still want to add that as a Concern.
- **How Extism works**: I don't like how Extism works, In Extism has "host functions" which then will be exported into the Plugins. I'm not a big fan of this concept and this makes adding Plugin support much harder than if when relying on other system (e.g. ABI).

### Other Options
We could use native libs 🤷‍♂️. Im free for other ideas

**Comment by circuitsacul** [comment]
Personal opinion: Use wasmtime directly instead, relying on wit-bindgen for API.

Practically, users will be limited to a select few languages (rust, c, maybe go, maybe python, maybe javascript). Non-compiled languages (go, but especially python & javascript) will result in massive plugin sizes because they have to include the interpreter. Though iirc wasmtime is working on a solution to this to share interpreters.

Realistically though, this is going to be true not matter what you use, because someone is going to have to maintain the bindings for it to be ergonomic to use.

So, instead of thinking of wasm as "it'll support every language in existence", think of wasm as "we can support two languages simultaneously, and if we don't like our choices we can easily add a different one later."

Also, wit is a much better solution (imo) than extism's host functions & guest functions: https://github.com/bytecodealliance/wit-bindgen

Basically, you have a (set of) wit file(s) that describe
1. the functions that the host exposes to guests (plugins)
2. the functions that the guest is expected to expose to the host
3. common, language-agnostic types

Then, plugins would use this wit-file in conjunction with a wit-bindgen implementation in their own language to auto-generate bindings to the host (see the section on [components](https://github.com/bytecodealliance/wit-bindgen?tab=readme-ov-file#creating-a-component))

**Comment by Alvsch** [comment]
I've looked into lua a bit and it seems like good integration for a possible plugin system. [mlua](https://crates.io/crates/mlua)

Though the biggest problem I can think of would be how multiple lua script would contrast to the plugin "drag and drop" that bukkit/spigot uses. With WASM we wouldn't have that problem as all files are bundled together.

**Comment by ReCore-sys** [comment]
I made some considerable headway into JVM based plugins on FerrumC and it's very promising. 2 way function calls are quite easy and full JVM ecosystem support is dead simple. And since plugins are generally quite small, the memory overhead is actually tiny. We also wouldn't be limited to Java, since the progress I made was entirely in Kotlin, but any JVM language would work fine.

**Comment by PiterWeb** [comment]
Maybe plugins can be implemented like binaries (ex: .exe) or dinamic libraries (.dll, .so).

The pro of the dll is that they can expose different functions and they are smaller than binaries. 

The functions behaviour can be easy implemented in the binaries if you use something like ENV variables to send the required function to run to the executable file.

The plugins will be launched as a new process from the Pumpkin instance and the communication can be handled through stdin/stdout buffers. This way near every programming language can be used for plugins, the problem is that you may need different plugin file for every OS.

**Comment by Snowiiii** [comment]
Im now sure to not use Extism

**Comment by circuitsacul** [comment]
> Im now sure to not use Extism

Do you have plans on what you will use?

**Comment by Snowiiii** [comment]
> > Im now sure to not use Extism
> 
> Do you have plans on what you will use?

Currently not 100% sure, But i would like to have Rust as the programming language for plugins

**Comment by nyabinary** [comment]
What does FerrumMC plan to use?

**Comment by nyabinary** [comment]
What does FerrumC plan to use?

**Comment by DataM0del** [comment]
> Extism doubles our overall dependency count and makes our `Cargo.lock` so much bigger

...then add `Cargo.lock` to the gitignore like I and some other people do???

**Comment by Alvsch** [comment]
> ...then add `Cargo.lock` to the gitignore like I and some other people do???

Already the case. The problem is really long compile times.

**Comment by 4lve** [comment]
> I've looked into lua a bit and it seems like good integration for a possible plugin system. [mlua](https://crates.io/crates/mlua)
> 
> 
> 
> Though the biggest problem I can think of would be how multiple lua script would contrast to the plugin "drag and drop" that bukkit/spigot uses. With WASM we wouldn't have that problem as all files are bundled together.

I think lua is a really good choice. The one file problem can be solved by either bundling the lua or zipping all the scripts
